lib/kdl/include/kdl/string_compare_detail.h

```diff
@@ -0,0 +1,224 @@
+/*
+ Copyright 2010-2019 Kristian Duske
+
+ Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+ documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
+ rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+ persons to whom the Software is furnished to do so, subject to the following conditions:
+
+ The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+ Software.
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+ WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+#ifndef TRENCHBROOM_STRING_COMPARE_DETAIL_H
+#define TRENCHBROOM_STRING_COMPARE_DETAIL_H
+
+#include <algorithm> // for std::mismatch, std::sort, std::search, std::equal
+#include <string_view>
+
+#include "collection_utils.h"
+
+namespace kdl {
+    /**
+     * Returns the first position at which the given strings differ. Characters are compared for equality using the
+     * given binary predicate.
+     *
+     * If the given strings are identical, then the length of the strings is returned.
+     *
+     * @tparam CharEqual the type of the binary predicate used to test characters for equality
+     * @param s1 the first string
+     * @param s2 the second string
+     * @param char_equal the binary predicate
+     * @return the first position at which the given strings differ
+     */
+    template <typename CharEqual>
```

Comment:
I would lean towards the string utils being non-header-only.

e.g. the comparators could be function pointers:
```
using CharEqual = bool (*)(char, char);
using CharCompare = int (*)(char, char);
```

Reply:
as discussed, it makes no difference in build times